### PR TITLE
feat: execute() exposes ExecutionReport

### PR DESCRIPTION
**Pull request by mattstam**
allows callers can check cycles, also removes cycle check / simulation from proof requests

### CODE REVIEW

File: core/src/runtime/mod.rs [hunk 1]
```diff
@@ -82,25 +83,44 @@ pub struct Runtime {
...
 
-impl InstructionReport {
-    pub fn total_instruction_count(&self) -> u32 {
+impl ExecutionReport {
+    pub fn total_instruction_count(&self) -> u64 {
         self.instruction_counts.values().sum()
     }
 }
 
+impl Display for ExecutionReport {
```
Comment (mattstam) [inline core/src/runtime/mod.rs:105]: also adding formatting

```
ExecutionReport { instruction_counts: {JALR: 49, BEQ: 26, BNE: 529, LBU: 109, SLL: 679, ADD: 4143, SRL: 1192, MUL: 502, SB: 23, SUB: 514, MULH: 500, SLTU: 10, BGEU: 16, SRA: 500, JAL: 14, XOR: 706, DIVU: 1, BLTU: 28, AUIPC: 28, AND: 249, LW: 392, BLT: 1, OR: 692, SW: 366, ECALL: 22}, syscall_counts: {HALT: 1, HINT_LEN: 1, HINT_READ: 1, COMMIT_DEFERRED_PROOFS: 8, WRITE: 3, COMMIT: 8} }
```

File: core/src/runtime/mod.rs [hunk 2]
```diff
@@ -82,25 +83,44 @@ pub struct Runtime {
...
     /// Whether we should write to the report.
     pub should_report: bool,
 }
 
 #[derive(Default, Debug, Clone, PartialEq, Eq)]
-pub struct InstructionReport {
-    pub instruction_counts: HashMap<Opcode, u32>,
-    pub syscall_counts: HashMap<SyscallCode, u32>,
+pub struct ExecutionReport {
+    pub instruction_counts: HashMap<Opcode, u64>,
```
Comment (mattstam) [inline core/src/runtime/mod.rs:95]: bump from u32 as we are already close to limit

File: core/src/runtime/mod.rs [hunk 3]
```diff
@@ -82,25 +83,44 @@ pub struct Runtime {
...
+        for (opcode, count) in &self.instruction_counts {
+            write!(f, "  {}: {}\n", opcode, count)?;
+        }
+        write!(
+            f,
+            "Total Instructions: {}\n",
+            self.total_instruction_count()
+        )?;
+        write!(f, "Syscall Counts:\n")?;
+        for (syscall, count) in &self.syscall_counts {
```
Comment (ctian1) [inline core/src/runtime/mod.rs:117]: it'd be nice to sort this in alphabetical order so its consistent

**Review by ctian1 — CHANGES_REQUESTED** [review]
I realized wherever we add to the report, we also need to check if unconstrained mode is on

File: core/src/runtime/mod.rs (hunk 3 above)
Comment (mattstam) [reply core/src/runtime/mod.rs]: updated

Comment (mattstam) [reply core/src/runtime/mod.rs]: Instruction Counts:
  add: 583439
  and: 151701
  auipc: 19488
  beq: 26917
  bge: 237
  bgeu: 5917
  blt: 1141
  bltu: 43457
  bne: 51442
  divu: 6
  ecall: 2264
  jal: 5372
  jalr: 38749
  lb: 10723
  lbu: 57950
  lw: 237094
  mul: 4036
  mulhu: 1152
  or: 301944
  sb: 58448
  sll: 278698
  sltu: 16766
  srl: 293010
  sub: 12382
  sw: 312781
  xor: 242242
Total Instructions: 2757356
Syscall Counts:
  16: 8
  26: 8
  0: 1
  6: 1091
  5: 1091
  2: 65
Total Syscall Count: 2264

Comment (mattstam) [reply core/src/runtime/mod.rs]: the difference in lower vs upper case is annoying, is it fine if I re-write all the Opcode enums https://github.com/succinctlabs/sp1/blob/dbed499f7aadbacc28a8e6e91a8761c5e7481d7d/core/src/runtime/opcode.rs#L72 to uppercase?

Comment (ctian1) [reply core/src/runtime/mod.rs]: can you just do .to_uppercase()?

Comment (ctian1) [reply core/src/runtime/mod.rs]: for consistency, should we show 0 for the missing ones?

**Review by ctian1 — APPROVED** [review]

**Pull request merged**